### PR TITLE
Cumulus 718 kinesis integration tests

### PR DESCRIPTION
**Pull request by flamingbear**
**Summary:** Summary of changes

Addresses [CUMULUS-718: Integration test that demonstrates Cumulus triggering a workflow configured to respond to a message from a Kinesis stream](https://bugs.earthdata.nasa.gov/browse/CUMULUS-718)

## Changes

* Adds two new specs that set up and tear down a kinesis stream for testing. (We can discuss if later this should be pulled out into CF)
* Adds kinesisTest spec that drops a Cloud Notification Message on a running kinesis stream and waits for the work flow to complete.  It tests that PODAAC's CNMToCMA lambda function performs as expected. And that syncGranules completes.
* Adds a collection, provider, and rule are all to the test data setup
* Fake test data is added to @cumulus/test-data
* Adds new workflow to be triggered KinesisTriggerTest 

## Test Plan
Things that should succeed before merging.

- [x] Unit tests
- [x] Adhoc testing
- [x] Update CHANGELOG
- [ ] Run `./bin/eslint-ratchet` and verify that eslint errors have not increased
  - [ ] Commit `.eslint-ratchet-high-water-mark` if the score has improved
- [x] ensure CNMToCMA lambda is properly stored in cumulus-shared-data bucket.
- [ ] run integration tests with "[it]" 

Suggested Reviewers: 
@laurenfrederick @abarciauskas-bgse @jennyhliu

### CODE REVIEW

File: example/spec/helpers/kinesisHelpers.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+const testConfig = loadConfig();
+
+const lambdaStep = new LambdaStep();
+const sfn = new StepFunctions({ region: testConfig.awsRegion });
+const kinesis = new Kinesis({ apiVersion: '2013-12-02', region: testConfig.awsRegion });
+
+const waitPeriodMs = 1000;
+
+/**
+ * returns the most recently executed KngestTriggerTest workflow
```
Comment (laurenfrederick) [inline example/spec/helpers/kinesisHelpers.js]: KinesisTriggerTest

File: example/spec/helpers/kinesisHelpers.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  let lastExecution;
+  let workflowExecution;
+
+  /* eslint-disable no-await-in-loop */
+  while (timeWaited < maxWaitTime && workflowExecution === undefined) {
+    try {
+      await timeout(waitPeriodMs);
+      timeWaited += waitPeriodMs;
+      try {
+        lastExecution = await getLastExecution();
```
[the remark this answers is not in the document]
Comment (flamingbear) [reply example/spec/helpers/kinesisHelpers.js]: Totally, this was an early cleanup I just never looked at again.

[the remark this answers is not in the document]
Comment (flamingbear) [reply example/spec/helpers/kinesisHelpers.js]: Actually now that I'm looking at it, do we really even need to log the error.  I think I was logging all those things when I was trying to figure out what the original code did.  What if I took them all out? 
although I just had an error running that step of the tests.  what about d56584ed

File: packages/integration-tests/index.js [hunk 3]
```diff
@@ -258,8 +258,8 @@ async function addProviders(stackName, bucketName, dataDirectory, s3Host = null)
  * @returns {Promise.<integer>} number of rules added
  */
 async function addRules(config, dataDirectory) {
-  const { stackName, bucketName } = config;
-  const rules = await setupSeedData(stackName, bucketName, dataDirectory);
+  const { stackName, bucket } = config;
```
Comment (Jkovarik) [inline packages/integration-tests/index.js]: I think we should probably make this consistent for all bucketName/bucket variables that at their root come from the testUtils.loadConfig  (config.bucket) for consistency (e.g. in addCollection, addProvider), and probably make the signatures consistent between this method and  addProviders/addCollections as well.

File: example/spec/helpers/kinesisHelpers.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+  timeout
+} = require('@cumulus/integration-tests');
+
+const { loadConfig } = require('../helpers/testUtils');
+
+const testConfig = loadConfig();
+
+const lambdaStep = new LambdaStep();
+const sfn = new StepFunctions({ region: testConfig.awsRegion });
+const kinesis = new Kinesis({ apiVersion: '2013-12-02', region: testConfig.awsRegion });
```
Comment (flamingbear) [inline example/spec/helpers/kinesisHelpers.js]: @laurenfrederick Should these be pushed into `@cumulus/common/aws`.  I thought of that yesterday when marc was walking us though his code.

File: packages/integration-tests/index.js (hunk 3 above)
Comment (flamingbear) [reply packages/integration-tests/index.js]: Yeah,  def for the signature.  I think I had good reasons for the bucket/bucketName when I did it, but bucketName matches the rest of this file.  I'll change that too.

Comment (flamingbear) [reply packages/integration-tests/index.js]: Well, actually.  The config is used to template the rules.json files because we need to replace AWS region and accountID before uploading to the DB.  So I think we need to pass in the whole config for adding rules.  

Weirdly, there's pseudoTemplating because addProviders overrides the `host` key only in the special case of an S3 protocol . https://github.com/nasa/cumulus/pull/455/files#diff-c758a8c9441f858914c13f40f7b9a94fR244 .  

I don't know if it's possible or necessary to unify the signatures and templating of the json files..

File: example/data/rules/L2_HR_PIXC_kinesisRule.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  "collection": {
+    "name": "L2_HR_PIXC",
+    "version": "000"
+  },
+  "name": "L2_HR_PIXC_kinesisRule",
+  "provider": "PODAAC_SWOT",
+  "rule": {
+    "type": "kinesis",
+    "value": "arn:aws:kinesis:{{awsRegion}}:{{awsAccountId}}:stream/{{streamName}}",
+    "arn": "arn:aws:kinesis:{{awsRegion}}:{{awsAccountId}}:stream/{{streamName}}"
```
Comment (jennyhliu) [inline example/data/rules/L2_HR_PIXC_kinesisRule.json]: rule.value is the kinesis stream that kinesis-consumer lambda listens to.
rule.arn is the EventSourceMappings id which is automatically added when the mapping is created.
The rule needs to be created after the kinesis stream is in place.  I don't see this in the PR.

File: packages/integration-tests/index.js (hunk 3 above)
Comment (Jkovarik) [reply packages/integration-tests/index.js]: 👍

File: packages/api/models/rules.js [hunk 6]
```diff
@@ -159,16 +159,25 @@ class Rule extends Manager {
           return (mapping.EventSourceArn === item.rule.value);
         });
       if (mappingExists) {
-        item.rule.arn = mappingExists.UUID;
-        return item;
+        const mappingEnabled = listData.EventSourceMappings
```
Comment (jennyhliu) [inline packages/api/models/rules.js]: We created the event source mapping with state 'ENABLED'.  Do we know how the state becomes 'not Enabled'?

File: packages/api/models/rules.js [hunk 7]
```diff
@@ -159,16 +159,25 @@ class Rule extends Manager {
...
+          .find((mapping) => { // eslint-disable-line arrow-body-style
+            return (mapping.EventSourceArn === item.rule.value &&
+                    mapping.State === 'Enabled');
+          });
+
+        if (mappingEnabled) {
+          item.rule.arn = mappingEnabled.UUID;
+          return item;
+        }
+        await aws.lambda().deleteEventSourceMapping({ UUID: mappingExists.UUID }).promise();
```
Comment (jennyhliu) [inline packages/api/models/rules.js]: The mapping could be shared by other rules, use rules.deleteKinesisEventSource() method instead.

Comment (flamingbear) [reply packages/api/models/rules.js]: It's in a new file. The stream is created at the beginning of integration tests, and destroyed at the end.  
https://github.com/nasa/cumulus/pull/455/files#diff-f4a144d837fc6c0abc86b61b63973879R10

File: example/data/rules/L2_HR_PIXC_kinesisRule.json (hunk 5 above)
Comment (flamingbear) [reply example/data/rules/L2_HR_PIXC_kinesisRule.json]: It's in a new file. The stream is created at the beginning of integration tests, and destroyed at the end.  
https://github.com/nasa/cumulus/pull/455/files#diff-f4a144d837fc6c0abc86b61b63973879R10

File: example/spec/helpers/kinesisHelpers.js (hunk 4 above)
Comment (abarciauskas-bgse) [reply example/spec/helpers/kinesisHelpers.js]: I think all of these kinesis helpers should go in @cumulus/common/aws - but would be fine if we added a ticket to do that later.

File: packages/api/models/rules.js (hunk 7 above)
Comment (jennyhliu) [reply packages/api/models/rules.js]: This api/models/rules.js is part of the api, not only for integration tests.

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+  granules: [
+    {
+      granuleId: granuleId,
+      files: [fileDataWithFilename]
+    }
+  ]
+};
+
+
+// When Cumulus is configured to trigger a CNM workflow from a Kinesis stream and a message appears on the stream, Cumulus triggers the workflow
```
Comment (abarciauskas-bgse) [inline example/spec/kinesisTests/KinesisTestTriggerSpec.js]: It might be helpful to be more specific, e.g.:
> When kinesis-type rules exist, the Cumulus lambda `kinesisConsumer` is configured to trigger workflows when new records arrive on a Kinesis stream. When a record appears on the stream, the `kinesisConsumer` lambda triggers workflows associated with the kinesis-type rules.

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+    await createOrUseTestStream(streamName);
+  });
+
+  it('waits for the stream to be active.', async () => {
+    await waitForActiveStream(streamName);
+  });
+
+  it('places a record on the stream.', async () => {
+    await putRecordOnStream(streamName, record);
+  });
```
Comment (abarciauskas-bgse) [inline example/spec/kinesisTests/KinesisTestTriggerSpec.js]: I think these three tests `finds or creates a test stream`, `waits for the stream to be active` and `places a record on the stream` should be in a beforeAll block.

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    expect(this.workflowExecution).not.toBe(undefined);
+  });
+
+  it('waits for step function to complete', async () => {
+    executionStatus = await waitForCompletedExecution(this.workflowExecution.executionArn);
+  });
+
+  it('executes successfully', () => {
+    expect(executionStatus).toEqual('SUCCEEDED');
+  });
```
Comment (abarciauskas-bgse) [inline example/spec/kinesisTests/KinesisTestTriggerSpec.js]: I think most of tests `The Cloud Notification Mechanism Kinesis workflow` should be in a beforeAll block - most are not testing anything. Only `finds a valid workflow execution.` and `executes successfully` are tests.

Comment (abarciauskas-bgse) [reply example/spec/kinesisTests/KinesisTestTriggerSpec.js]: Same with some other tests in this PR, e.g. `Closes a kinesis test stream`

File: packages/api/models/rules.js [hunk 11]
```diff
@@ -159,16 +159,25 @@ class Rule extends Manager {
...
       }
     }
 
     // create event source mapping
     const params = {
       EventSourceArn: item.rule.value,
       FunctionName: process.env.kinesisConsumer,
-      StartingPosition: 'LATEST',
+      StartingPosition: 'TRIM_HORIZON',
```
Comment (abarciauskas-bgse) [inline packages/api/models/rules.js]: What's the reason for this change?

File: packages/integration-tests/index.js [hunk 12]
```diff
@@ -271,6 +271,43 @@ async function addRules(config, dataDirectory) {
...
+/**
+ *
+ * @param {string} stackName - Cloud formation stack name
+ * @param {string} bucketName - S3 internal bucket name
+ * @param {Array} rules - List of rules objects to delete
+ * @returns {Promise.<integer>} - Number of rules deleted
+ */
+async function deleteRules(stackName, bucketName, rules) {
+  setProcessEnvironment(stackName, bucketName);
+  const promises = rules.map((rule) => limit(() => _deleteOneRule(rule.name)));
```
Comment (abarciauskas-bgse) [inline packages/integration-tests/index.js]: Is there a reason we can't use https://github.com/nasa/cumulus/blob/master/packages/api/models/rules.js#L35-L50 ?

**Comment by abarciauskas-bgse** [comment]
@flamingbear you probably know this but this branch is out of date with master :)

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js (hunk 10 above)
Comment (flamingbear) [reply example/spec/kinesisTests/KinesisTestTriggerSpec.js]: So is that actually bad?  it was a style thing for me.  Originally all of the setup was in the beforeAll, and when something failed it just thew it's exception and then started running the tests and I had a really hard time figuring out where the error was coming from.

I like that that each step in the setup was it's own step that could succeed or fail and that the steps tell you what's happening in real time as your watching them.    But if this is bad form, I can move them back into a before all.  (But, I really didn't like the way after beforeAll failed, the tests just tried to run anyway and just fail with errors unrelated to the problem that occurred.)

File: packages/api/models/rules.js (hunk 11 above)
Comment (flamingbear) [reply packages/api/models/rules.js]: This was the hardest thing to figure out. When I added a stream setup and teardown. Even after ensuring the stream was 'ACTIVE' and the eventSourceMapping was in place and 'ENABLED', the first message on the stream never triggered the kinesisConsumer.   

I struggled trying to figure out what needed to be in place, and originally set this to be `StartingPosition: 'AT_TIMESTAMP'` with a timestamp of now()  which kind of made sense to me. That would mean that any message placed on the stream after the eventmapping was created would be processed, but alireza said it should be TRIM_HORIZON to process everything on the stream.

File: packages/integration-tests/index.js (hunk 12 above)
Comment (flamingbear) [reply packages/integration-tests/index.js]: Do you mean get rid of _deleteOneRule() helper?  I think that's what that is doing. 
https://github.com/nasa/cumulus/blob/CUMULUS-718-kinesis-integration-tests/packages/integration-tests/index.js#L274-L283

**Comment by flamingbear** [comment]
@abarciauskas-bgse Of course it's out of date, we're two days from end of sprint. :) I'l fix that too.

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js (hunk 8 above)
Comment (flamingbear) [reply example/spec/kinesisTests/KinesisTestTriggerSpec.js]: That's funny because I was going to delete that comment, I copied it from the jira ticket since patrick said to make our tests follow that language.

Comment (flamingbear) [reply example/spec/kinesisTests/KinesisTestTriggerSpec.js]: but I will update it with your text.

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js (hunk 10 above)
Comment (abarciauskas-bgse) [reply example/spec/kinesisTests/KinesisTestTriggerSpec.js]: I know what you mean - my preferred adaptation would be to make sure the test suite exits if the beforeAll block doesn't complete successfully. I looked into this a bit and it looks like jasmine doesn't offer a `fail fast` option (ref: https://stackoverflow.com/questions/11562817/does-jasmine-node-offer-any-type-of-fail-fast-option, which links to a bunch of related jasmine github issues). I would be interested to know if the `process.exit(1)` solution works.

Alternatively, if we keep these as unit tests, we should actually assert something - what do you think?

File: packages/api/models/rules.js (hunk 11 above)
Comment (abarciauskas-bgse) [reply packages/api/models/rules.js]: Hmm I wonder if this some sort of bug in the way AWS configures the lambda trigger when starting position is `LATEST` - like it doesn't know what "LATEST" is until there is at least one record having been on the stream. But `TRIM_HORIZON` makes sense.

File: example/spec/kinesisTests/KinesisTestTriggerSpec.js (hunk 10 above)
Comment (flamingbear) [reply example/spec/kinesisTests/KinesisTestTriggerSpec.js]: Ok, if we leave as is, you'd like to see `expects().something()` in each  `it` "test".  Or you'd like to see all of the non-test setup in a beforeAll()?  
I bet you have a preference for the latter?  and have to go look to see if I prefer the first.

File: packages/integration-tests/index.js (hunk 12 above)
Comment (abarciauskas-bgse) [reply packages/integration-tests/index.js]: I see, I thought these were already rules objects. My mistake, you can ignore this comment.

File: packages/api/models/rules.js (hunk 11 above)
Comment (flamingbear) [reply packages/api/models/rules.js]: " like it doesn't know what "LATEST" is until there is at least one record having been on the stream" . I wondered this exactly but never got around to testing it.

File: packages/integration-tests/index.js (hunk 3 above)
Comment (flamingbear) [reply packages/integration-tests/index.js]: Actually we can't pull bucketName out of config, and need the bucket object from config.

File: example/spec/helpers/kinesisHelpers.js (hunk 4 above)
Comment (laurenfrederick) [reply example/spec/helpers/kinesisHelpers.js]: Yeah that is an excellent point 👍 If you want to merge to get unblocked for the next ticket, you could do it there as well.

File: packages/api/models/rules.js (hunk 6 above)
Comment (flamingbear) [reply packages/api/models/rules.js]: Ok, I was able to reproduce the state:
```     { UUID: '5f0dc194-5cab-4149-a292-e408689779a9',
       BatchSize: 100,
       EventSourceArn: 'arn:aws:kinesis:us-east-1:893015583569:stream/mhs-cumulus-testStream',
       FunctionArn: 'arn:aws:lambda:us-east-1:893015583569:function:mhs-cumulus-kinesisConsumer',
       LastModified: 2018-07-19T19:38:00.000Z,
       LastProcessingResult: 'PROBLEM: Stream not found. Recreate the stream and re-enable the event source mapping to resume processing from TRIM_HORIZON.',
       State: 'Disabled',
       StateTransitionReason: 'Unrecoverable error' },
```

File: packages/integration-tests/index.js (hunk 3 above)
Comment (Jkovarik) [reply packages/integration-tests/index.js]: Hmm, indeed.  We could reassign it to be ... more clear, but that doesn't really feel useful.       The issue (for posterity) is the key in the config doesn't match what we're calling it in the code because the other functions  are ... translating, whereas addRules is pulling the value directly.     This is a level of pedantry that isn't helpful in this PR, but I might take a stab at a refactor elsewhere - I'll drop the concern here.    Thanks for thinking about it.

File: example/app/config.yml [hunk 13]
```diff
@@ -244,3 +247,51 @@ jl:
...
+    shared:
+      name: cumulus-data-shared
+      type: shared
+  api_distribution_url: '{{API_DISTRIBUTION_URL}}'
+  system_bucket: '{{buckets.internal.name}}'
+
+  users:
+    - username: '{{EARTHDATA_LOGIN_USERNAME}}'
+
+  AutoScalingPolicyConfiguration:
```
Comment (abarciauskas-bgse) [inline example/app/config.yml]: Does this key need to be here?

Comment (flamingbear) [reply example/app/config.yml]: That's in my personal config and we're out of the sandbox, so it turns off autoscaling for us at NSIDC.  More preventative than anything I think.

**Comment by flamingbear** [comment]
I'd like to clean up the getStepOutput, but I'd like to not have this open anymore even more.  @laurenfrederick if you want to talk about what that function should do and if it should be cleaned up some.
